### PR TITLE
including links in newState

### PR DESCRIPTION
**Pull request by tomharvey**
To demonstrate a potential fix for issue 100 https://github.com/dixieio/redux-json-api/issues/100 - adding access to pagination links as a non-breaking update.

Adds links to the newState object when reading from the API. 
I'm Testing for this in `link-payload.js` using the fixture in `linked_payload.json`.

Apologies for the chaining of the `addLinksToState` and `updateOrInsertResourcesIntoState` methods; I'm old/new to the JS scene and not sure what the best style is these days.

### CODE REVIEW

**Comment by egeriis** [comment]
Hey @tomharvey 

Thank you for contributing.

While I can see that this may provide some access through state, to the links. I don't see the broad use for doing it like this. Since links are stored in `api.links`, they are easily untrustable if you're doing more than 1 API request that responds with links. How do you view this?

**Comment by tomharvey** [comment]
Good point. I do think it's important to allow access to the top level `links`; there is a use case when receiving data from a paginated API. Although you are correct in spotting the potential for unexpected behaviour. I'll try to justify an alternative solution.

Let's consider the following response from a single API endpoint:

``` json
GET /posts
{
  "data": [
    {
      "id": "1",
      "type": "posts",
      "attributes": {
        "title": "A post"
      },
      "links": {
        "self": "http://example.com/post/1"
      }
    },
    {
      "id": "2",
      "type": "posts",
      "attributes": {
        "title": "Another post"
      },
      "links": {
        "self": "http://example.com/post/2"
      }
    }
  ],
  "links": {
    "self": "http://example.com/posts",
    "next": "http://example.com/posts?page[offset]=2",
    "last": "http://example.com/posts?page[offset]=10"
  }
}
```

According to the jsonapi spec:

* `links` on the top level object are related to the primary `data`.
* Each object is capable of having a `links` object as a member.

The above example demonstrates a potential API response where`links` are used for both of these use cases.

In the current state of this PR the `api` object would contain the following objects:

```
api.posts.links
>>> {"self": "http://example.com/post/1"}
```

and

```
api.links
>>> {
	 "self": "http://example.com/posts",
    "next": "http://example.com/posts?page[offset]=2",
    "last": "http://example.com/posts?page[offset]=10"
  }
```

You correctly point out that if we made a secondary call, for example, to read a paginated list of all authors on the blogging team then `api.links` would then be ambiguous in its contents; either containing the links for pages of `authors` or `posts`.

I think that we should __explicitly__ attribute the top level `links` to the primary data object of the response, in line with the jsonapi spec. But we must avoid conflicting with the distinct, existing case of a list of links as a child each of the primary data objects - demonstrated above as `api.posts.links`.

Perhaps the best solution for this would be to have a top level `links` object in the `api` object but make the list of links a child of the primary data object type, so:

```
api.links
>>> {
  "posts": [
    "http://example.com/posts",
    "http://example.com/posts?page[offset]=2",
    "http://example.com/posts?page[offset]=10"
  ],
  "seperately-requested-data-type": [
    "http://example.com/seperately-requested-data-type",
    "http://example.com/seperately-requested-data-type?page[offset]=2",
    "http://example.com/seperately-requested-data-type?page[offset]=10"
  ]
}
```

or the client could use

```
api.links.posts
>>> [
  "http://example.com/posts",
  "http://example.com/posts?page[offset]=2",
  "http://example.com/posts?page[offset]=10"
  ]
```

to access the top level `links` object returned when calling the `posts`

Let me know if you're happy with the above suggestions and I'll work up a change to the PR.

Thanks for your feedback.

**Comment by tomharvey** [comment]
So, I made the change to the structure of links I suggested above - quiet end to the week, so I thought I'd go ahead.

Looking forward to your feedback, again my ES6 isn't great so sorry for any poor syntax or patterns.

**Comment by egeriis** [comment]
Hey @tomharvey,

Appreciate your revision and your reply.

I think it is a good idea to move forward with your proposal. I am a bit strict in my approach to implementing these kind of features, that are not 100% clear to the user. I can see a couple of use cases where the user of _redux-json-api_ would get confused with the altered approach too.

While I am not a fan of this approach, we are already talking about adding a configuration object to `readEndpoint` in another PR thread: #108 
We could utilize a similar approach for this feature, by adding a "indexLinks" configuration property, that when set would trigger the storage of the responded links on a specified property.

Let me provide you with an example:

```js
readEndpoint('/users?filter[name]=john', { indexLinks: 'usersSearch' });
```

My idea is that the `links` from response body would then be put into `api.links.usersSearch`.

As far as I can see, this would eliminate any confusion about stored links and potential conflict that would arise from any assumptions implemented.

What do you think?

**Comment by tomharvey** [comment]
Sounds good 👍 Let me work that up.

I had started to think that allowing the developer to pass some kind of namespace with the readendpoint would allow for better separation of how the library handles multiple calls. I'm still a little unsure about the way that `GET /posts` and `GET /posts/1` will still return one `api.posts` object. Certainly not unhappy enough to complain, just to wonder if there was a need for this kind of namespacing in other areas beyond the links.

Thanks

**Comment by tomharvey** [comment]
@egeriis  to avoid conflicting with the exciting https://github.com/dixieio/redux-json-api/pull/108 I've tried to avoid changing too much in `jsonapi.js` as I see that this PR is in that file as well. Otherwise, tests and implementation for what I understand you were suggesting.

File: test/link-payload.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+import { reducer } from '../src/jsonapi';
+
+const apiRead = createAction('API_READ');
+
+describe('[State mutation] Insertion of links', () => {
+  it('should not insert links into state when indexLinks is blank', () => {
+    const state = {};
+
+    apiRead(postsPayload);
```
Comment (egeriis) [inline test/link-payload.js]: This line doesn't do anything. Is it supposed to be there?

File: src/jsonapi.js [hunk 2]
```diff
@@ -149,7 +151,7 @@ export const readEndpoint = (endpoint, {
         credentials: 'include'
       })
         .then(json => {
-          dispatch(apiRead({ endpoint, ...json }));
+          dispatch(apiRead({ endpoint, indexLinks, ...json }));
```
Comment (egeriis) [inline src/jsonapi.js]: Is it necessary to proxy `indexLinks` through the payload?

**Comment by tomharvey** [comment]
@egeriis Agreed, let's not pollute the api. Did you mean something like this? Thanks

File: src/state-mutation.js [hunk 3]
```diff
@@ -85,14 +85,14 @@ const stateContainsResource = (state, resource) => {
...
+export const addLinksToState = (state, links, options) => {
+  if (options === undefined || options.indexLinks === undefined) {
     return state;
   }
   const newState = state;
 
   const typedLinks = {};
-  typedLinks[linkName] = links;
+  typedLinks[options.indexLinks] = links;
   newState.links = Object.assign({}, state.links, typedLinks);
```
Comment (egeriis) [inline src/state-mutation.js]: We have a neat library called `object-path-immutable` included. Using that here would make it more readable :) https://github.com/mariocasciaro/object-path-immutable

**Comment by egeriis** [comment]
Looks great now @tomharvey :)

**Comment by tomharvey** [comment]
Thanks @egeriis that's a nice little library, good to know about it. I've implemented it now.

**Comment by tomharvey** [comment]
@egeriis Let me know if there's anything more you need on this to get it merged. Once it's merged I'll build on the work @mansona did on https://github.com/dixieio/redux-json-api/pull/108 and use the same approach as here.

**Pull request merged**